File: 백준/Gold/1967. 트리의 지름/트리의 지름.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**9)

n = int(input())
tree = [[] for _ in range(n + 1)]
for _ in range(n - 1):
    a, b, w = map(int, input().split())
    tree[a].append((b, w))
    tree[b].append((a, w))

maxnode = 0
maxlen = 0
visited = [False] * (n + 1)
# 부모 노드 i에서 가장 멀리 나온 노드
def dfs(i, l):
    global maxlen
    global maxnode
    for t in tree[i]:
        if not visited[t[0]]:
            visited[t[0]] = True
            dfs(t[0], l + t[1])
            visited[t[0]] = False
    else:
        if l > maxlen:
            maxlen = l
            maxnode = i

visited[1] = True
dfs(1, 0)

# maxnode에서 가장 멀리 나온 노드
visited = [False] * (n + 1)
visited[maxnode] = True
dfs(maxnode, 0)
print(maxlen)

File: 백준/Gold/1967. 트리의 지름/트리의 지름.java
import java.io.*;
import java.util.*;

public class Main {
    static boolean[] visited;
    static ArrayList<Integer[]>[] tree;
    static int depth = 0;
    static int target = 1;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        tree = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            tree[i] = new ArrayList<>();
        }
        for (int i = 0; i < n - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int parent = Integer.parseInt(st.nextToken());
            int child = Integer.parseInt(st.nextToken());
            int length = Integer.parseInt(st.nextToken());
            tree[parent].add(new Integer[] {child, length});
            tree[child].add(new Integer[] {parent, length});
        }

        // 루트에서 가장 먼 곳
        visited = new boolean[n + 1];
        visited[1] = true;
        dfs(1, 0);

        // 가장 먼 곳에서 또 가장 먼 곳
        visited = new boolean[n + 1];
        visited[target] = true;
        depth = 0;
        dfs(target, 0);

        System.out.println(depth);
    }

    static void dfs(int n, int d) {
        // 자식들
        for (Integer[] to : tree[n]) {
            if (!visited[to[0]]) {
                visited[to[0]] = true;
                if (depth < d + to[1]) {
                    target = to[0];
                    depth = d + to[1];
                }
                dfs(to[0], d + to[1]);
            }
        }
    }

}